src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java

```diff
@@ -142,4 +144,122 @@ public void close() throws IOException {
+    }
+
+    private static long supplyUsingBlockingChannel(Channel c, ThrowingLongSupplier<IOException> tls) throws IOException {
+        if (c instanceof SelectableChannel sc) {
+            synchronized (sc.blockingLock()) {
+                if (!sc.isBlocking()) {
+                    throw new IllegalBlockingModeException();
+                }
+                return tls.supply();
+            }
+        } else {
+            return tls.supply();
+        }
+    }
+
+    private static long doTransfer(ReadableByteChannel rbc, WritableByteChannel wbc) throws IOException {
+        if (rbc instanceof FileChannel fc) {
+            return transfer(fc, wbc);
+        }
+
+        if (wbc instanceof FileChannel fc) {
+            if (rbc instanceof SeekableByteChannel sbc) {
+                return transfer(sbc, fc);
+            }
+
+            return transfer(rbc, fc);
+        }
+
+        return transfer(rbc, wbc);
+    }
+
+    private static long transfer(FileChannel src, WritableByteChannel dst) throws IOException {
+        long bytesWritten = 0L;
+        long srcPos = src.position();
+        try {
+            while (srcPos + bytesWritten < src.size()) {
+                bytesWritten += src.transferTo(srcPos + bytesWritten, Long.MAX_VALUE, dst);
+            }
+            return bytesWritten;
+        } finally {
+            src.position(srcPos + bytesWritten);
+        }
+    }
+
+    private static long transfer(SeekableByteChannel src, FileChannel dst) throws IOException {
+        long bytesWritten = 0L;
+        long dstPos = dst.position();
+        long srcPos = src.position();
+        try {
+            while (srcPos + bytesWritten < src.size()) {
+                bytesWritten += dst.transferFrom(src, dstPos + bytesWritten, Long.MAX_VALUE);
+            }
+            return bytesWritten;
+        } finally {
+            dst.position(dstPos + bytesWritten);
+        }
+    }
+
+    private static long transfer(ReadableByteChannel src, FileChannel dst) throws IOException {
+        long bytesWritten = 0L;
+        long dstPos = dst.position();
+        ByteBuffer bb = Util.getTemporaryDirectBuffer(TRANSFER_SIZE);
+        try {
+            int bytesRead;
+            do {
+                bytesWritten += dst.transferFrom(src, dstPos + bytesWritten, Long.MAX_VALUE);
+                bytesRead = src.read(bb); // detect end-of-stream
+                if (bytesRead > -1) {
+                    bb.flip();
+                    while (bb.hasRemaining())
+                        dst.write(bb);
+                    bb.clear();
+                    bytesWritten += bytesRead;
+                }
+            } while (bytesRead > -1);
+            return bytesWritten;
+        } finally {
+            dst.position(dstPos + bytesWritten);
+            Util.releaseTemporaryDirectBuffer(bb);
+        }
+    }
+
+    private static long transfer(ReadableByteChannel src, WritableByteChannel dst) throws IOException {
```

Comment:
Does this method have a measurable improvement in performance over `InputStream.transferTo()`?

Reply:
Not really, and did actually not expect much, as the overhead of ChannelInputStream.read / ChannelOutputStream.write only is that ByteBuffer.wrap() is performed repeatedly in a loop, while my code uses the same utility buffer always. Nevertheless, my code reduces GC pressure as it does not create objects or java byte arrays on the heap in a loop, so the more iterations the more GC reduction. Overall I would say, my code is beneficial even if the *speed* does not improve much.

Reply:
Measured it today, here are the actual results:
```
Sandbox.transferTo  thrpt   25  0,091 ± 0,011  ops/s -- using byte[] (original implementation)
Sandbox.transferTo  thrpt   25  0,095 ± 0,012  ops/s -- using utility temporary buffer (my contribution)
Sandbox.transferTo  thrpt   23  0,099 ± 0,012  ops/s -- FileChannel.transferTo(FileChannel)
```

The improvement of this method actually is approx. 4% on my Windows laptop, which I would say is worth to do it.